integrations/azure-devops/azure_devops/client.py

```diff
@@ -0,0 +1,268 @@
+import json
+from typing import Any, AsyncGenerator, Optional
+from azure_devops.webhooks.webhook_event import WebhookEvent
+from port_ocean.context.event import event
+from port_ocean.context.ocean import ocean
+from loguru import logger
+from .http.http_client import HTTPBaseClient
+from azure_devops.utils import cache_results
+
+API_URL_PREFIX = "_apis"
+
+
+class AzureDevopsClient(HTTPBaseClient):
+    def __init__(self, organization_url: str, personal_access_token: str) -> None:
+        super().__init__(personal_access_token)
+        self._organization_base_url = organization_url
+
+    @classmethod
+    def create_from_ocean_config(cls) -> "AzureDevopsClient":
+        if cache := event.attributes.get("azure_devops_client"):
+            return cache
+        azure_devops_client = cls(
+            ocean.integration_config["organization_url"],
+            ocean.integration_config["personal_access_token"],
+        )
+        event.attributes["azure_devops_client"] = azure_devops_client
+        return azure_devops_client
+
+    @cache_results("projects")
+    async def generate_projects(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        params = {"includeCapabilities": "true"}
+        projects_url = f"{self._organization_base_url}/{API_URL_PREFIX}/projects"
+        async for projects in self._get_paginated_by_top_and_continuation_token(
+            projects_url, additional_params=params
+        ):
+            yield projects
+
+    @cache_results("teams")
+    async def generate_teams(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        teams_url = f"{self._organization_base_url}/{API_URL_PREFIX}/teams"
+        async for teams in self._get_paginated_by_top_and_skip(teams_url):
+            yield teams
+
+    async def generate_members(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        async for teams in self.generate_teams():
+            for team in teams:
```

Comment:
- i dont think that this function responsability is also to fetch the teams. The one calling it should pass team to it
- each team will be executed synchronically, try to use `asyncio.as_completed`